### PR TITLE
[R-package] factored dependency 'magrittr' out of R package

### PR DESCRIPTION
**Pull request by jameslamb**
@Laurae2 in this PR, I propose that we completely factor out the dependency `magrittr`, in the interest of reducing `lightgbm`'s dependency footprint. I know `magrittr` itself has minimal additional dependencies, but in general I think we should strive to keep `lightgbm` as light as possible and I feel this particular dependency isn't providing a lot of value.

### CODE REVIEW

**Comment by jameslamb** [comment]
Just rebased this to `master` to keep up with recent changes.

File: R-package/R/lgb.importance.R [hunk 1]
```diff
@@ -44,22 +43,33 @@ lgb.importance <- function(model, percentage = TRUE) {
...
 
   # Extract elements
-  tree_imp <- tree_dt %>%
-    magrittr::extract(.,
-                      i = ! is.na(split_index),
-                      j = .(Gain = sum(split_gain), Cover = sum(internal_count), Frequency = .N),
-                      by = "split_feature") %T>%
-    data.table::setnames(., old = "split_feature", new = "Feature") %>%
-    magrittr::extract(., i = order(Gain, decreasing = TRUE))
+  tree_imp_dt <- tree_dt[
```
Comment (Laurae2) [inline R-package/R/lgb.importance.R]: comma should at beginning of lines or at end of lines by convention?

I think we should standardize them in the package, likely to be at beginning of lines as it's easier/clearer to read.

Comment (jameslamb) [reply R-package/R/lgb.importance.R]: I believe it should be beginning of lines for R code. Would you like me to open an issue and make a PR to standardize?

**Comment by StrikerRUS** [comment]
@jameslamb Can we merge this? If so, I'll be able to remove `magrittr` dependency from #2176, which is going to be merged this weekend.

**Comment by jameslamb** [comment]
> @jameslamb Can we merge this? If so, I'll be able to remove `magrittr` dependency from #2176, which is going to be merged this weekend.

Seems that one file is in conflict. Let me rebase and run the tests manually (because I haven't done #2335  yet).

**Comment by jameslamb** [comment]
Ok we currently have 4 failing tests:

```
══ testthat results  ═══════════════════════════════════════════════════════════
[ OK: 33 | SKIPPED: 0 | WARNINGS: 3 | FAILED: 4 ]
1. Error: training continuation works (@test_basic.R#71)
2. Failure: Feature penalties work properly (@test_parameters.R#36)
3. Failure: Feature penalties work properly (@test_parameters.R#37)
4. Failure: Feature penalties work properly (@test_parameters.R#38)
```

The "training continuation" one exists on `master`, but I think the other three are problems I've introduced in this PR. I don't feel comfortable merging until I address those three. WIll try to do it today.

File: R-package/R/lgb.importance.R (hunk 1 above)
Comment (jameslamb) [reply R-package/R/lgb.importance.R]: opened #2373 and added PR #2437 to address it

File: R-package/R/lgb.model.dt.tree.R [hunk 2]
```diff
@@ -64,10 +63,16 @@ lgb.model.dt.tree <- function(model, num_iteration = NULL) {
...
+  # Substitute feature index with the actual feature name
+
+  # Since the index comes from C++ (which is 0-indexed), be sure
+  # to add 1 (e.g. index 28 means the 29th feature in feature_names)
+  split_feature_indx <- tree_dt[, split_feature] + 1
+
+  # Get corresponding feature names. Positions in split_feature_indx
+  # which are NA will result in an NA feature name
+  feature_names <- parsed_json_model$feature_names[split_feature_indx]
+  tree_dt[, split_feaure := feature_names]
```
Comment (jameslamb) [inline R-package/R/lgb.model.dt.tree.R]: WOW I think this is why the tests were failing for me on this build. typo: `split_feaure` 😆 

testing now

**Comment by jameslamb** [comment]
> Ok we currently have 4 failing tests:
> 
> ```
> ══ testthat results  ═══════════════════════════════════════════════════════════
> [ OK: 33 | SKIPPED: 0 | WARNINGS: 3 | FAILED: 4 ]
> 1. Error: training continuation works (@test_basic.R#71)
> 2. Failure: Feature penalties work properly (@test_parameters.R#36)
> 3. Failure: Feature penalties work properly (@test_parameters.R#37)
> 4. Failure: Feature penalties work properly (@test_parameters.R#38)
> ```
> 
> The "training continuation" one exists on `master`, but I think the other three are problems I've introduced in this PR. I don't feel comfortable merging until I address those three. WIll try to do it today.

Alright I've figured out the three test errors I introduced. R unit tests now all work. I need to fix the "training continuation works" one but this PR doesn't need to wait on that.

@Laurae2 could you take one more look at this PR and let me know if I could merge?

**Comment by StrikerRUS** [comment]
@jameslamb Please remove this dependency from here as well.
https://github.com/microsoft/LightGBM/blob/e77658205d3b1bc5c19e4ba3821f4222f5be8c05/docs/conf.py#L232

**Comment by jameslamb** [comment]
> @jameslamb Please remove this dependency from here as well.
> https://github.com/microsoft/LightGBM/blob/e77658205d3b1bc5c19e4ba3821f4222f5be8c05/docs/conf.py#L232

done. Good catch, I forgot we do that. I did a `git grep magrittr` just to be sure there are no other uses in non-standard places, did not find any.

**Comment by StrikerRUS** [comment]
@jameslamb 
> done. Good catch, I forgot we do that. I did a git grep magrittr just to be sure there are no other uses in non-standard places, did not find any.

Thanks!

**Comment by jameslamb** [comment]
I just rebased this PR to `master`.

Tests for the R package are passing locally (see below)

```
export CXX=/usr/local/bin/g++-8 CC=/usr/local/bin/gcc-8
Rscript build_r.R
cd R-package/tests
Rscript testthat.R
```

Once all CI checks pass, I'm going to merge this.

**Pull request merged**